### ISSUE TITLE
Decide on graph walking vs EL inference for using taxon constraints to make subsets

### ISSUE THREAD

**Issue by cmungall** [obophenotype/uberon#2137]
Current strategy to make a taxon subset:

 - Add axiom Thing subClassOf part-of some NCBITaxon:nnnn
 - Remove any inter-species existentials
      - homology represented as reciprocal existentials
      - inter-species edges (less relevant for uberon)
 - Ensure all TCs are EL-ified
     - never_in becomes disjointWith in-taxon some X
     - Ensure NCBITaxon GCIs are added (in-taxon some X disjoint with in-taxon some Y for all sibs)
 - Reason with Elk
 - Eliminate all unsatisfiables

Note this has issues if we have:

 * A subclass part-of some B
 * B never-in-taxon

See 
https://github.com/geneontology/go-annotation/issues/3942

@balhoff says he has a solution

Some additional issues with the approach

 - hard to mentally reason over - logic is separated across ad-hoc filters on OPs, various pre-processing steps, property chains stored centrally in GO
 - Can require 10s of gigs of memory
      - this gets worse the further away from human we go
 - Hard to 'customize'. E.g. GO may want to include occurs_in when making their subsets

Outline alternative strategies in this ticket

**Comment by matentzn** [comment]
@dosumis if you want @shawntanzk, @anitacaron and me to act on this, we need some specific instructions as it will fall to @anitacaron to take the bulk of this work, and it will occupy her for a few weeks (given she only has a few hours a week to dedicate to this project I mean). 

From the meeting I gather we need to:

- [ ] Decide who should document, and where the documentation lives
- [ ] Decide on the curation strategy (how do we get the taxon constraints into the ontologies)
- [ ] Decide on the technical strategy to materialise the logical constraints (DOSDP vs SPARQL)
- [ ] Decide on the technical strategy on extracting taxon views on the basis of these constraints

@shawntanzk I think we can handle this after all, its just going to be a slow process. If you want, you can put it up on the board again for next week.

**Comment by cmungall** [comment]
For the graph strategy, it is easy to explore this using the existing ubergraph instance, which includes relation-graph inferences

See this query:
https://api.triplydb.com/s/8hs8rvxj3

Which is hardcoded to return classes EXCLUDED from a human view

Scroll up for an explanation of the query

Note that for demonstrative purposes, this is highly aggressive. For example, annotation shortcuts like spatially-disjoint-with are treated like any other triples in relation-graph (we should exclude not owl entailed NG in query). If there were homology assertions in any ontology, these would also be propagated over. However, it is trivial to exclude these either at sparql time or as a post-processing step

**Comment by cmungall** [comment]
Results of running on uberon-edit from Nov 1 here:

https://s3.amazonaws.com/bbop-ontologies/uberon/tmp/class-taxon-exclusions.tsv.gz

URL not guaranteed stable, this is for testing

The query is hardcodes for Human and Mammal. The TSV lists classes that are excluded for a given taxon, together with the reason.

Apologies for the duplication due to different labels of OPs:

|?c|?cLabel|?p|?pLabel|?clsWithConstraint|?clsWithConstraintLabel|?taxonWithConstraint|?taxonWithConstraintLabel|?queryTaxon|
|---|---|---|---|---|---|---|---|---|
|<http://purl.obolibrary.org/obo/UBERON_0003221>|"phalanx"|<http://purl.obolibrary.org/obo/RO_0002202>|"develops_from"|<http://purl.obolibrary.org/obo/UBERON_2001544>|"sublingual cartilage"|<http://purl.obolibrary.org/obo/NCBITaxon_40674>|"Mammalia"|<http://purl.obolibrary.org/obo/NCBITaxon_9606>|
|<http://purl.obolibrary.org/obo/UBERON_0003221>|"phalanx"|<http://purl.obolibrary.org/obo/RO_0002202>|"develops from"@en|<http://purl.obolibrary.org/obo/UBERON_2001544>|"sublingual cartilage"|<http://purl.obolibrary.org/obo/NCBITaxon_40674>|"Mammalia"|<http://purl.obolibrary.org/obo/NCBITaxon_9606>|
|<http://purl.obolibrary.org/obo/UBERON_0003221>|"phalanx"|<http://purl.obolibrary.org/obo/RO_0002202>|"develops from"|<http://purl.obolibrary.org/obo/UBERON_2001544>|"sublingual cartilage"|<http://purl.obolibrary.org/obo/NCBITaxon_40674>|"Mammalia"|<http://purl.obolibrary.org/obo/NCBITaxon_9606>|

This is obviously a false positve, caused by #2159 

Note: would be great to have robot output saner TSVs, can anyone work on:
https://github.com/ontodev/robot/issues/176

**Comment by shawntanzk** [comment]
@dosumis - could you provide some guidance for how the tech team can proceed with this? thanks

**Comment by matentzn** [comment]
Should be reconsidered eventually

**Comment by anitacaron** [comment]
Since last year, this has been a low priority. If it should have a higher priority, please give some action items.

**Comment by matentzn** [comment]
I think this is covered by the new subset command, we should double check

**Comment by gouttegd** [comment]
I am not sure this issue is relevant anymore. We have _already_ overhauled the way we are making taxon subsets in #3363, and so many points in the above discussion are no longer accurate.

We currently have two distinct strategies available to make taxon subsets (we can switch from one to another by changing one variable in the custom Makefile; the one we are currently using is `default`).

## “Default” strategy

This is basically the first strategy outlined by Chris in the initial comment: for a given root class _R_ (which could be `owl:Thing`, but we are rather using UBERON:0000000 and UBERON:0001062) and the taxon _T_, we assert _R SubClassOf: in_taxon some T_ and exclude all classes that becomes unsatisfiable as a result of that assertion.

Regarding the “additional issues“ mentioned by Chris about that strategy:

> hard to mentally reason over - logic is separated across ad-hoc filters on OPs, various pre-processing steps, property chains stored centrally in GO

This may have been true in the past. I disagree that it is the case now. The logic is very simple and entirely implemented in a single Java class in the Uberon ROBOT plugin. There is no “preprocessing“ required. Property chains are now mostly in RO.

> Can require 10s of gigs of memory

No. This does not require any larger amount of memory than what is already required for a typical Uberon build.

On my machine, building a taxon subset requires ~7GB of RAM and takes ~1min.

The one real problem with that strategy is that it may exclude more classes than it should, because _any relationship with an excluded class causes the linked classes to be excluded as well_, regardless of what the property is (in particular, regardless if there is a chain between that property and `in_taxon`). The second strategy has been designed precisely to avoid that problem.

## “Precise“ strategy

This is kind of the “opposite“ of the previous approach, in that we select the classes that are valid for a given taxon instead of excluding the classes that are invalid for that taxon.

For a given root class _R_ and the taxon _T_, we iterate over all subclasses _C_ of _R_ and include the subclasses for which the expression _C and in_taxon some T_ is satisfiable.

Compared to the previous approach, it is slower (about ~4 to ~5 times slower on my machine, about ~5 minutes to build one taxon subset). It will never excludes more classes than it should, since it only excludes a class that is linked to another excluded class iff the two classes are linked by a property over which `in_taxon` holds; however, it may include more classes than it should if RO is missing some property chains over `in_taxon`.

To summarise: the “default” strategy tends to exclude more classes than it should because it completely ignores (does not need) the property chains over `in_taxon`; the “precise” strategy tends to include more classes than it should because it relies on the completeness of the set of property chains over `in_taxon`.

As I said, we can switch from one strategy to another by changing one line in the Makefile. It’s unclear to me why we should need yet another strategy.
